pkg/controller/studycontroller/studycontroller_controller.go

```diff
@@ -0,0 +1,480 @@
+	}
+	go r.controllerloop(instance)
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileStudyController) getStudyConf(instance *katibv1alpha1.StudyController) (*katibapi.StudyConfig, error) {
+
+	sconf := &katibapi.StudyConfig{
+		Metrics: []string{},
+		ParameterConfigs: &katibapi.StudyConfig_ParameterConfigs{
+			Configs: []*katibapi.ParameterConfig{},
+		},
+	}
+	if instance.Spec.StudySpec.Name == "" {
+		return nil, fmt.Errorf("StudyName must be set")
+	}
+
+	sconf.Name = instance.Spec.StudySpec.Name
+	sconf.Owner = instance.Spec.StudySpec.Owner
+	if instance.Spec.StudySpec.OptimizationGoal != nil {
+
+		sconf.OptimizationGoal = *instance.Spec.StudySpec.OptimizationGoal
+	}
+	sconf.ObjectiveValueName = instance.Spec.StudySpec.ObjectiveValueName
+	switch instance.Spec.StudySpec.OptimizationType {
+	case katibv1alpha1.OptimizationTypeMinimize:
+		sconf.OptimizationType = katibapi.OptimizationType_MINIMIZE
+	case katibv1alpha1.OptimizationTypeMaximize:
+		sconf.OptimizationType = katibapi.OptimizationType_MAXIMIZE
+	default:
+		sconf.OptimizationType = katibapi.OptimizationType_UNKNOWN_OPTIMIZATION
+	}
+	for _, m := range instance.Spec.StudySpec.MetricsNames {
+		sconf.Metrics = append(sconf.Metrics, m)
+	}
+	for _, pc := range instance.Spec.StudySpec.ParameterConfigs {
+		p := &katibapi.ParameterConfig{
+			Feasible: &katibapi.FeasibleSpace{},
+		}
+		p.Name = pc.Name
+		p.Feasible.Min = pc.Feasible.Min
+		p.Feasible.Max = pc.Feasible.Max
+		p.Feasible.List = pc.Feasible.List
+		switch pc.ParameterType {
+		case katibv1alpha1.ParameterTypeUnknown:
+			p.ParameterType = katibapi.ParameterType_UNKNOWN_TYPE
+		case katibv1alpha1.ParameterTypeDouble:
+			p.ParameterType = katibapi.ParameterType_DOUBLE
+		case katibv1alpha1.ParameterTypeInt:
+			p.ParameterType = katibapi.ParameterType_INT
+		case katibv1alpha1.ParameterTypeDiscrete:
+			p.ParameterType = katibapi.ParameterType_DISCRETE
+		case katibv1alpha1.ParameterTypeCategorical:
+			p.ParameterType = katibapi.ParameterType_CATEGORICAL
+		}
+		sconf.ParameterConfigs.Configs = append(sconf.ParameterConfigs.Configs, p)
+	}
+	return sconf, nil
+}
+
+func (r *ReconcileStudyController) checkGoal(instance *katibv1alpha1.StudyController, sconf *katibapi.StudyConfig, mr *katibapi.GetMetricsReply) bool {
+	if instance.Spec.StudySpec.OptimizationGoal == nil {
+		return false
+	}
+	for _, mls := range mr.MetricsLogSets {
+		for _, ml := range mls.MetricsLogs {
+			if ml.Name == sconf.ObjectiveValueName {
+				curValue, _ := strconv.ParseFloat(ml.Values[len(ml.Values)-1], 32)
+				if sconf.OptimizationType == katibapi.OptimizationType_MINIMIZE && curValue < sconf.OptimizationGoal {
+					return true
+				} else if sconf.OptimizationType == katibapi.OptimizationType_MAXIMIZE && curValue > sconf.OptimizationGoal {
+					return true
+				} else {
+					return false
+				}
+			}
+		}
+	}
+	return false
+}
+
+func (r *ReconcileStudyController) controllerloop(instance *katibv1alpha1.StudyController) {
```

Comment:
I think we could add some comments for the func, since it is the main control loop.